pkg/storage/compress.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2020 PingCAP, Inc. Licensed under Apache-2.0.
+
+package storage
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+
+	"github.com/pingcap/errors"
+)
+
+type withCompression struct {
+	ExternalStorage
+	compressType CompressType
+}
+
+// UnwrapCompression unwraps withCompression to original ExternalStorage
+func UnwrapCompression(storage ExternalStorage) ExternalStorage {
+	if compressExt, ok := storage.(*withCompression); ok {
+		return UnwrapCompression(compressExt.ExternalStorage)
+	}
+	return storage
+}
+
+// WithCompression returns an ExternalStorage with compress option
+func WithCompression(inner ExternalStorage, compressionType CompressType) ExternalStorage {
+	if compressionType == NoCompression {
+		return UnwrapCompression(inner)
+	}
+	return &withCompression{ExternalStorage: UnwrapCompression(inner), compressType: compressionType}
+}
```

Comment:
`WithCompression` should *add* compression to an external storage, not replace it. If we make `WithCompression(WithCompression(inner, Gzip), Gzip)` it should creates a storage that compresses twice (no matter how silly it looks).